logger/modules/logger.py

```diff
@@ -0,0 +1,153 @@
+
+            if len(entries) == 0:
+                print("ERROR: The directory for this log session was not found.")
+                return False, None
+
+            log_names = [
+                entry for entry in entries if os.path.isdir(os.path.join(log_directory_path, entry))
+            ]
+
+            # Find the log directory for the current run, which is the most recent timestamp.
+            log_path = max(
+                log_names,
+                key=lambda datetime_string: datetime.datetime.strptime(
+                    datetime_string, file_datetime_format
+                ),
+            )
+
+            filepath = pathlib.Path(log_directory_path, log_path, f"{name}.log")
+
+            file_handler = logging.FileHandler(filename=filepath, mode="w")
+            file_handler.setFormatter(formatter)
+            logger.addHandler(file_handler)
+
+        return True, Logger(cls.__create_key, logger)
+
+    def __init__(self, class_create_private_key: object, logger: logging.Logger) -> None:
+        """
+        Private constructor, use create() method.
+        """
+        assert class_create_private_key is Logger.__create_key, "Use create() method."
+
+        self.logger = logger
+
+    @staticmethod
+    def message_and_metadata(message: str, frame: "types.FrameType | None") -> str:
+        """
+        Extracts metadata from frame and appends it to the message.
+        """
+        if frame is None:
+            return message
+
+        # Get Pylance to stop complaining
+        assert frame is not None
+
+        function_name = frame.f_code.co_name
+        filename = frame.f_code.co_filename
+        line_number = inspect.getframeinfo(frame).lineno
+
+        return f"[{filename} | {function_name} | {line_number}] {message}"
+
+    def debug(self, message: str, frame: "types.FrameType | None") -> None:
+        """
+        Logs a debug level message.
+        """
+        message = self.message_and_metadata(message, frame)
+        self.logger.debug(message)
+
+    def info(self, message: str, frame: "types.FrameType | None") -> None:
+        """
+        Logs an info level message.
+        """
+        message = self.message_and_metadata(message, frame)
+        self.logger.info(message)
+
+    def warning(self, message: str, frame: "types.FrameType | None") -> None:
+        """
+        Logs a warning level message.
+        """
+        message = self.message_and_metadata(message, frame)
+        self.logger.warning(message)
+
+    def error(self, message: str, frame: "types.FrameType | None") -> None:
+        """
+        Logs an error level message.
+        """
+        message = self.message_and_metadata(message, frame)
+        self.logger.error(message)
+
+    def critical(self, message: str, frame: "types.FrameType | None") -> None:
+        """
+        Logs a critical level message.
+        """
+        message = self.message_and_metadata(message, frame)
+        self.logger.critical(message)
```

Comment:
I propose that rather than asking the user to pass in a frame, we use inspect in the Logger itself, and we can use [frame.f_back](https://docs.python.org/3/library/inspect.html#types-and-members) to find the caller's frame, which is the one we are interested in. Then we can just have frame be a boolean option that defaults to true, since it seems to be enabled by default on most stack traces and loggers.